sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,591 @@
+from sympy import acos
+from sympy.physics.vector import cross, Vector, dot
+from sympy.physics.mechanics import dynamicsymbols
+
+__all__ = ['Joint', 'PinJoint', 'SlidingJoint', 'CylindricalJoint',
+           'SphericalJoint', 'PlanarJoint']
+
+
+class Joint(object):
+    """Base class for all specific joints.
+
+    Joints connects two bodies (a parent and child) by adding degrees of freedom
+    to child w.r.t parent. Different joints adds different degrees of freedom.
+    This is the base class for all specific joints and holds all common methods
+    acting as an interface for all joints. Custom joint can also be created
+    by creating a subclass of this class and overriding 'apply_joint()'
+    method to add the dynamics of the joint.
+
+    Parameters
+    ----------
+    name: String
+        Name of the joint which makes it unique. Should be different from other
+        joints.
+    parent: Body
+        Instance of Body class serving as the parent in the joint.
+    child: Body
+        Instance of Body class serving as the child in the joint.
+    parent_point_pos: 3 Tuple, Optional
+        Defines the joint's point where the parent will be connected to child.
+        3 Tuple defines the values of x, y and z directions w.r.t parent's
+        frame. If it is not supplied, center of mass is added as default.
+    child_point_pos: 3 Tuple, Optional
+        Defines the joint's point where the child will be connected to parent.
+        3 Tuple defines the values of x, y and z directions w.r.t child's frame.
+        If it is not supplied, center of mass is added as default.
+    parent_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with child's
+        axis before adding joint. If it is not passed, default is x axis in
+        parent's frame.
+    child_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with parent's
+        axis before adding joint. If it is not passed, default is x axis in
+        child's frame.
+    """
+
+    def __init__(self, name, parent, child, parent_point_pos=None,
+                 child_point_pos=None, parent_axis=None, child_axis=None):
```

Comment:
It isn't apparent to me that all of these kwargs are applicable to all types of joints. For example, a spherical/ball joint would only need a point.